File: solutions/prolog/space-age/space_age.pl
conversion("Earth", 1.0).
conversion("Mercury", 0.2408467).
conversion("Venus", 0.61519726).
conversion("Mars", 1.8808158).
conversion("Jupiter", 11.862615).
conversion("Saturn", 29.447498).
conversion("Uranus", 84.016846).
conversion("Neptune", 164.79132).

seconds_to_years(Seconds, Years) :-
    Years is Seconds / 31557600.

space_age(Planet, AgeSec, Years) :-
    conversion(Planet, YearFactor),
    seconds_to_years(AgeSec, AgeYears),
    Years is AgeYears / YearFactor.

File: solutions/fsharp/space-age/SpaceAge.fs
module SpaceAge

type Planet =
    | Earth  | Mercury | Venus
    | Mars   | Jupiter | Saturn
    | Uranus | Neptune 
    
let earthYearLength = 31557600.0
    
let age (planet: Planet) (seconds: int64) : float =
    let conversionFactor =
        match planet with
            | Earth   -> 1.0
            | Mercury -> 0.2408467
            | Venus   -> 0.61519726
            | Mars    -> 1.8808158
            | Jupiter -> 11.862615
            | Saturn  -> 29.447498
            | Uranus  -> 84.016846
            | Neptune -> 164.79132
    (float)seconds / earthYearLength / conversionFactor
